### PR TITLE
Add Provisioning ability for IBM Cloud Self-managed

### PR DESCRIPTION
**Pull request by bdattoma**
extending the provisionin/deprovisioning tasks to work with IBM Cloud

### CODE REVIEW

File: ods_ci/tasks/Tasks/provision_self_managed_cluster.robot [hunk 1]
```diff
@@ -16,17 +16,20 @@ ${hive_kubeconf}         %{KUBECONFIG}
...
 ${artifacts_dir}         ${OUTPUT DIR}
 
+
 *** Tasks ***
 Provision Self-Managed Cluster
     [Documentation]    Provision a self-managed cluster
     [Tags]  self_managed_provision
     Provision Cluster
-    Claim Cluster
+    IF    ${use_cluster_pool}    Claim Cluster
```
Comment (bdattoma) [inline ods_ci/tasks/Tasks/provision_self_managed_cluster.robot:32]: to fix, this line and line 19 will fail

**Comment by bdattoma** [comment]
IBM Cloud provisioning/deprovisioning tested from local multiple times: PASSED. Cannot test on CI yet because of missing pkgs

Other cloud provider regression testing provisioning/deprovisionin: _coming soon_

**Review by manosnoam — APPROVED** [review]
Thank you Berto, looks good! Can you add a run output/build number ?
Also, I see several linter warnings, if you can cleanup.

File: ods_ci/tasks/Resources/Provisioning/Hive/provision.robot [hunk 2]
```diff
@@ -145,31 +193,43 @@ Watch Hive Install Log
     Should Contain    ${new_log_data}    install completed successfully
 
 Wait For Cluster To Be Ready
-    ${pool_namespace} =    Get Cluster Pool Namespace    ${pool_name}
-    Set Task Variable    ${pool_namespace}
-    Log    Watching Hive Pool namespace: ${pool_namespace}    console=True
+    IF    ${use_cluster_pool}
+        ${pool_namespace} =    Get Cluster Pool Namespace    ${pool_name}
+        Set Task Variable    ${pool_namespace}
+        Set Task Variable    ${clusterdeployment_name}    ${pool_namespace}
```
Comment (manosnoam) [inline ods_ci/tasks/Resources/Provisioning/Hive/provision.robot:199]: @bdattoma FYI:
- The Cluster deployment name has a similar name (if I recall) as the cluster pool namespace by default.
- The Cluster pool namespace is a unique name, which includes the cluster name prefix.
- We should not make the cluster name equal to cluster pool namespace (which is currently the case for GCP and AWS).

**Review by manosnoam — APPROVED** [review]
Besides the few questions/notes, lgtm

File: ods_ci/tasks/Resources/Provisioning/Hive/provision.robot (hunk 2 above)
Comment (bdattoma) [reply ods_ci/tasks/Resources/Provisioning/Hive/provision.robot:199]: I just replicated the logic that was there earlier.
The code was using `${pool_namespace}` variable as cluster deployment name when using cluster pools.

File: ods_ci/tasks/Resources/Provisioning/Hive/deprovision.robot [hunk 3]
```diff
@@ -12,27 +18,55 @@ Set Hive Default Variables
...
-    Log Many    @{Delete_Cluster}
+    IF    ${use_cluster_pool}
+        Log    Deleting cluster ${cluster_name} configuration    console=True
+        @{Delete_Cluster} =    Oc Delete    kind=ClusterPool    name=${pool_name}
+        ...    namespace=${hive_namespace}    api_version=hive.openshift.io/v1
+        Log Many    @{Delete_Cluster}
+        ${Delete_Cluster} =    Oc Delete    kind=ClusterDeploymentCustomization    name=${conf_name}
+        ...    namespace=${hive_namespace}    api_version=hive.openshift.io/v1
+        Log Many    @{Delete_Cluster}
+    ELSE
```
[the remark this answers is not in the document]
Comment (bdattoma) [reply ods_ci/tasks/Resources/Provisioning/Hive/deprovision.robot:29]: >Can you add Documentation info when is it required not to use cluster pool,

ok

>and when do we need to delete/keep the IBM resources ?

Didn't get, what do you mean? 
When deleting the cluster, if the provider is IBM it must delete the resources

[the remark this answers is not in the document]
Comment (bdattoma) [reply ods_ci/tasks/Resources/Provisioning/Hive/deprovision.robot:29]: I've added few lines in the TaskSuite documentation about Pools vs Deployment

[the remark this answers is not in the document]
Comment (manosnoam) [reply ods_ci/tasks/Resources/Provisioning/Hive/deprovision.robot:29]: I mean that by default ${use_cluster_pool} is set to True. When it is False it deletes more resources on IBM infra (afaik from the else block). When is it happen that we will need to set ${use_cluster_pool}=False ?

**Review by manosnoam — APPROVED** [review]

**Review by kobihk — COMMENTED** [review]
Thank you @bdattoma for the additional provision option.

I added 2 comments

File: ods_ci/tasks/Resources/Provisioning/Hive/provision.robot [hunk 4]
```diff
@@ -8,17 +10,37 @@ Claim Cluster
...
+            ...    api_version=hive.openshift.io/v1    fields=['spec.clusterName']
+            Log    ${clusterdeploymentname}
+            Log    ${cluster_name}
+            IF    "${clusterdeploymentname}" == "${cluster_name}"
+                RETURN    ${clusterdeploymentname}
+            END
+        ELSE
+            Log    message=No ClusterDeployment found in ${hive_namespace}.
+        END
+        RETURN    False
```
[the remark this answers is not in the document]
Comment (bdattoma) [reply ods_ci/tasks/Resources/Provisioning/Hive/provision.robot:42]: you're right, missed it while adding the ELSE logic. I'll do the change tomorrow

File: ods_ci/tasks/Resources/Provisioning/Hive/IBM/ibmcloud-cluster.yaml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          name: ibmcloud-creds
+    provisioning:
+      imageSetRef:
+        name: ${infrastructure_configurations}[image_set]
+      installConfigSecretRef:
+        name: ibmcloud-sno-install-config
+      manifestsSecretRef:
+        name: ${infrastructure_configurations}[hive_cluster_name]-manifests
+    # pullSecretRef:
+    #   name: ${infrastructure_configurations}[hive_cluster_name]-pull-secret
```
[the remark this answers is not in the document]
Comment (bdattoma) [reply ods_ci/tasks/Resources/Provisioning/Hive/IBM/ibmcloud-cluster.yaml]: done

**Review by kobihk — APPROVED** [review]

**Review by tarukumar — APPROVED** [review]

**Pull request merged**